# Boston Housing Assignment

Implimenting scikit learn's r2 score and mse methods for measuring the performance of the linear regressor and sklearn.linear_model.Ridge modle

In [1]:
from sklearn import datasets
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.cross_validation import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from sklearn.linear_model import LinearRegression

In [2]:
bean = datasets.load_boston()
print bean.DESCR

Boston House Prices dataset

Notes
------
Data Set Characteristics:  

    :Number of Instances: 506 

    :Number of Attributes: 13 numeric/categorical predictive
    
    :Median Value (attribute 14) is usually the target

    :Attribute Information (in order):
        - CRIM     per capita crime rate by town
        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
        - INDUS    proportion of non-retail business acres per town
        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
        - NOX      nitric oxides concentration (parts per 10 million)
        - RM       average number of rooms per dwelling
        - AGE      proportion of owner-occupied units built prior to 1940
        - DIS      weighted distances to five Boston employment centres
        - RAD      index of accessibility to radial highways
        - TAX      full-value property-tax rate per $10,000
        - PTRATIO  pupil-teacher ratio by town
      

In [3]:
def load_boston():
    scaler = StandardScaler()
    boston = datasets.load_boston()
    X=boston.data
    y=boston.target
    X = scaler.fit_transform(X)
    return train_test_split(X,y)

In [4]:
X_train, X_test, y_train, y_test = load_boston()

In [5]:
X_train.shape

(379L, 13L)

## Fitting a Linear Regression

This is done by instantiating a new regression object (line 1) and giving your regression object your training data (line 2) by calling .fit(independent variables, dependent variable)

In [6]:
clf = LinearRegression()
clf.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

## Making a Prediction

In [7]:
zip (y_test, clf.predict(X_test))

[(23.300000000000001, 28.50994326043061),
 (19.100000000000001, 16.382962001997399),
 (17.5, 16.32096762173925),
 (19.199999999999999, 20.314725458588899),
 (21.699999999999999, 20.636783969673672),
 (11.9, 22.990102165195566),
 (11.800000000000001, 9.162218759197053),
 (20.399999999999999, 23.170233292457713),
 (19.800000000000001, 18.648226901846979),
 (18.899999999999999, 15.056755307592818),
 (32.899999999999999, 30.371185543376313),
 (23.100000000000001, 25.143624013665175),
 (20.199999999999999, 16.288708945351942),
 (23.899999999999999, 28.170096756359307),
 (25.100000000000001, 29.387619677644203),
 (15.4, 15.207646397905885),
 (23.100000000000001, 10.777103770179558),
 (18.899999999999999, 19.051828913760076),
 (19.399999999999999, 17.442896342237027),
 (15.6, 19.143813240879204),
 (23.800000000000001, 23.201502647821542),
 (22.199999999999999, 22.453798328540248),
 (43.5, 39.645376438980151),
 (22.100000000000001, 27.320091526527975),
 (12.0, 11.059611594249185),
 (34.8999999

In [8]:
mean_squared_error(y_test,clf.predict(X_test))

21.947408595071991

In [9]:
r2_score(y_test,clf.predict(X_test))

0.72721145346968408

## MSE = 21.947408595071991

## r2_score = 0.72721145346968408

## Computing Linear Regression

taking alpha=0.001 for computing sklearn.linear_model.Ridge


In [10]:
from sklearn.linear_model import Ridge
r = Ridge(alpha=.001)
r.fit(X_train, y_train)

Ridge(alpha=0.001, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001)

In [11]:
zip (y_test, clf.predict(X_test))

[(23.300000000000001, 28.50994326043061),
 (19.100000000000001, 16.382962001997399),
 (17.5, 16.32096762173925),
 (19.199999999999999, 20.314725458588899),
 (21.699999999999999, 20.636783969673672),
 (11.9, 22.990102165195566),
 (11.800000000000001, 9.162218759197053),
 (20.399999999999999, 23.170233292457713),
 (19.800000000000001, 18.648226901846979),
 (18.899999999999999, 15.056755307592818),
 (32.899999999999999, 30.371185543376313),
 (23.100000000000001, 25.143624013665175),
 (20.199999999999999, 16.288708945351942),
 (23.899999999999999, 28.170096756359307),
 (25.100000000000001, 29.387619677644203),
 (15.4, 15.207646397905885),
 (23.100000000000001, 10.777103770179558),
 (18.899999999999999, 19.051828913760076),
 (19.399999999999999, 17.442896342237027),
 (15.6, 19.143813240879204),
 (23.800000000000001, 23.201502647821542),
 (22.199999999999999, 22.453798328540248),
 (43.5, 39.645376438980151),
 (22.100000000000001, 27.320091526527975),
 (12.0, 11.059611594249185),
 (34.8999999

In [12]:
mean_squared_error(y_test,r.predict(X_test))

21.947363350725592

In [13]:
r2_score(y_test,r.predict(X_test))

0.72721201582033412

# Ridge Linear Regression

## MSR = 21.947363350725592

## r2_score = 0.72721201582033412